### PR TITLE
Fix Screen Space ambient occlusion which was broken since a while

### PR DESCRIPTION
**Pull request by samcake**
This PR simply fixes Ambient Occlusion effect which was broken in the past few weeks.
Evolve the Resource mangement so that the effect is in charge of the framebuffers it need

Fixed several bugs that were there and giving wrong results.

Quality wise it s a bit better but there a re still many issues to fix.
THis is mostly trying to reenable it and bring it backto a working state

### CODE REVIEW

**Comment by samcake** [comment]
To test
IN GRaphics enbale the Ambient OCclusion effect
Everything should render correctly with a tiny bit of ao
IN desktop it should works quite correctly with probably some artefacts but not worse that what it was.
In HMD, the noise pattern is still flickering from frames to frames, i didn;t really fixed that. next time maybe...

**Comment by hifi-gustavo** [comment]
Build console output for commit: b90e0cfd898dc1315b8c76c15acd82a251975a3c available at: https://gist.github.com/c3a32b682a9d9abb40f09bb244db6825<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-b90e0cf.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-b90e0cf.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: b90e0cfd898dc1315b8c76c15acd82a251975a3c available at: https://gist.github.com/c3a32b682a9d9abb40f09bb244db6825<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-b90e0cf.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-b90e0cf.dmg

**Comment by Nex-Pro** [comment]
#### FAIL

IN GRaphics enbale the Ambient OCclusion effect **CHECK**
Everything should render correctly with a tiny bit of ao **FAIL**

Particles are broken. Particles are only rendered in front of some solid objec such as the moon at playa.

See attached image.

particleerror_8433

**Comment by Nex-Pro** [comment]
I carefully compared this against build #5244

**Comment by hifi-gustavo** [comment]
Build console output for commit: 87d61abe5aa261ec1050dcbf76fa088d48b4fea4 available at: https://gist.github.com/c3a32b682a9d9abb40f09bb244db6825<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-87d61ab.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-87d61ab.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: b5296410c3a928d42b115aa78d6cd7e563262f08 available at: https://gist.github.com/c3a32b682a9d9abb40f09bb244db6825<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-b529641.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-b529641.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6a076bcee4316d40bb7bec8c652c2abc4cff1145 available at: https://gist.github.com/c3a32b682a9d9abb40f09bb244db6825<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-6a076bc.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-6a076bc.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5ad66a17eeaeda3de18d787c475c21dbd5a28fb5 available at: https://gist.github.com/3288218c9c1231d9ef25cbca14e6c495<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-5ad66a1.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-5ad66a1.dmg

**Comment by samcake** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5ad66a17eeaeda3de18d787c475c21dbd5a28fb5 available at: https://gist.github.com/c3a32b682a9d9abb40f09bb244db6825<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-5ad66a1.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-5ad66a1.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9cdd07921373719fb3e6d6ce6b989e3b2fe9f4d3 available at: https://gist.github.com/c3a32b682a9d9abb40f09bb244db6825<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-9cdd079.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-9cdd079.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 557f898d9580fc7e72996ecb6c1b16c720eec143 available at: https://gist.github.com/c3a32b682a9d9abb40f09bb244db6825<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-557f898.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-557f898.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4097fcb0f67d1f0e7016179fc45101aa1a6c08cd available at: https://gist.github.com/c3a32b682a9d9abb40f09bb244db6825<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-4097fcb.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-4097fcb.dmg

**Comment by marko8904** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4097fcb0f67d1f0e7016179fc45101aa1a6c08cd available at: https://gist.github.com/c3a32b682a9d9abb40f09bb244db6825<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-4097fcb.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-4097fcb.dmg

**Comment by marko8904** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4097fcb0f67d1f0e7016179fc45101aa1a6c08cd available at: https://gist.github.com/c3a32b682a9d9abb40f09bb244db6825<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-4097fcb.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-4097fcb.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1e84f9dc90d0169adf8d15b45aff9da4fd0a476d available at: https://gist.github.com/c3a32b682a9d9abb40f09bb244db6825<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-1e84f9d.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-1e84f9d.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4da712fbdf9daaf609a51de0f72a8709ee32116f available at: https://gist.github.com/c3a32b682a9d9abb40f09bb244db6825<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-4da712f.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-4da712f.dmg

**Comment by kevinmthomasse** [comment]
per @cozza13 request to re-test

**Comment by kevinmthomasse** [comment]
## QA Test
**QA Tester**: KevinMThomas
**QA Pass**: Success
**Additional Test Results + Observations**: Enabled ```AO``` and it worked correctly and reviewed @Nex-Pro notes and tested particles and did not seem to have an issue now with the new build as it definitely appeared before.  HMD mode worked just as well.  Screen shot attached.
capture

File: libraries/render-utils/src/AmbientOcclusionEffect.cpp [hunk 1]
```diff
@@ -272,76 +333,53 @@ const gpu::PipelinePointer& AmbientOcclusionEffect::getVBlurPipeline() {
...
+    float tHeight = occlusionViewport.w / (float)framebufferSize.y;
 
-        for (int i = 0; i < 2; i++) {
-            // Compose the mono Eye space to Stereo clip space Projection Matrix
-            auto sideViewMat = projMats[i] * eyeViews[i];
-            _frameTransformBuffer.edit<FrameTransform>().projection[i] = sideViewMat;
-        }
 
-        _frameTransformBuffer.edit<FrameTransform>().stereoInfo = glm::vec4(1.0f, (float)(args->_viewport.z >> 1), 0.0f, 1.0f);
+    //_parametersBuffer->ditheringInfo.y += 0.25f;
```
Comment (sethalves) [inline libraries/render-utils/src/AmbientOcclusionEffect.cpp]: did you mean to leave this here?

File: libraries/render-utils/src/SurfaceGeometryPass.cpp [hunk 2]
```diff
@@ -71,14 +74,15 @@ void LinearDepthFramebuffer::allocate() {
     // For Linear Depth:
     _linearDepthTexture = gpu::TexturePointer(gpu::Texture::create2D(gpu::Element(gpu::SCALAR, gpu::FLOAT, gpu::RGB), width, height,
         gpu::Sampler(gpu::Sampler::FILTER_MIN_MAG_LINEAR_MIP_POINT)));
-   // _linearDepthTexture->autoGenerateMips(1);
+ //   _linearDepthTexture->autoGenerateMips(5);
```
Comment (sethalves) [inline libraries/render-utils/src/SurfaceGeometryPass.cpp]: commented code

File: libraries/render-utils/src/AmbientOcclusionEffect.h [hunk 3]
```diff
@@ -126,22 +154,92 @@ class AmbientOcclusionEffect {
...
+class DebugAmbientOcclusionConfig : public render::Job::Config {
+    Q_OBJECT
+
+    Q_PROPERTY(bool showCursorPixel MEMBER showCursorPixel NOTIFY dirty)
+    Q_PROPERTY(glm::vec2 debugCursorTexcoord MEMBER debugCursorTexcoord NOTIFY dirty)
+public:
+    DebugAmbientOcclusionConfig() : render::Job::Config(true) {}
+
+    bool showCursorPixel{ false };
+    glm::vec2 debugCursorTexcoord{ 0.5, 0.5 };
```
Comment (sethalves) [inline libraries/render-utils/src/AmbientOcclusionEffect.h]: 0.5f ?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9669234f9c0bcd3f4e40f48befab19632ba20738 available at: https://gist.github.com/c3a32b682a9d9abb40f09bb244db6825<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-9669234.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-9669234.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8cc65ea9653244dbbaae7bcbb66ff099a0cb2ab0 available at: https://gist.github.com/c3a32b682a9d9abb40f09bb244db6825<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-8cc65ea.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8433-8cc65ea.dmg

**Pull request merged**